dags/legacy_test/tests/pytorch/nightly/common.libsonnet

```diff
@@ -120,7 +120,36 @@ local volumes = import 'templates/volumes.libsonnet';
   },
   GpuMixin:: {
     local config = self,
-    imageTag+: '_cuda_11.8',
+    imageTag+: '_cuda_12.1',
+
+    entrypoint: [
+      'bash',
+      '-cxue',
+      |||
+        export PATH=/usr/local/nvidia/bin${PATH:+:${PATH}}
+        export LD_LIBRARY_PATH=/usr/local/nvidia/lib64:/usr/local/nvidia/lib${LD_LIBRARY_PATH:+:${LD_LIBRARY_PATH}}
+
+        nvidia-smi
+        pip3 uninstall -y torch torchvision
+        pip3 install --pre torch torchvision --index-url https://download.pytorch.org/whl/nightly/cpu
+
+        mkdir pytorch
+        wget https://github.com/pytorch/xla/archive/refs/heads/master.tar.gz -O - | tar xzf -
+        mv xla-master pytorch/xla
+
+        export PJRT_DEVICE=CUDA
+
+        # Run whatever is in `command` here
+        "${@:0}"
+      |||,
+    ],
+    command: [
+      'torchrun',
+      '--nnodes=%d' % config.accelerator.num_hosts,
+      '--node_rank=$(JOB_COMPLETION_INDEX)',
+      '--nproc_per_node=%d' % config.accelerator.count,
+      '--rdzv_endpoint=$(JOB_NAME)-0.headless-svc:12355',
```

Comment:
This is very cool. Thanks Will!

Reply:
Trying to learn more. What's this `headless`?

Reply:
headless-svc is the name of the kubernets service.

Reply:
"Headless" is a concept from kubernetes: https://kubernetes.io/docs/concepts/services-networking/service/#headless-services

It gives all pods that match the selector a subdomain under its name. Since the service name is `headless-svc`, matching pods (ie those that define the right label) with names like `abc-123` will be addressable as `abc-123.headless-svc`. This is relevant to multihost GPU jobs, since Kubernetes gives them predictable hostnames like `$JOB_NAME-0`, `$JOB_NAME-1`, etc. Attaching them to a headless service makes them accessible by a predictable DNS name.

Reply:
Thank you @will-cromar for the explanation!